File: lib/ncurses.ml
type window
type m_event = { x : int; y : int; bstate : int }

external _initscr : unit -> window = "_initscr" "_initscr"
external _cbreak : unit -> unit = "_cbreak" "_cbreak"
external _noecho : unit -> unit = "_noecho" "_noecho"
external _clear : unit -> unit = "_clear" "_clear"
external _refresh : unit -> unit = "_refresh" "_refresh"
external _endwin : unit -> unit = "_endwin" "_endwin"
external _move : int -> int -> unit = "_move" "_move"
external _get_mouse : unit -> m_event option = "_get_mouse" "_get_mouse"
external _getch : unit -> int option = "_getch" "_getch"
external _keypad : unit -> unit = "_keypad" "_keypad"
external _intrflush : unit -> unit = "_intrflush" "_intrflush"
external _nodelay : unit -> unit = "_nodelay" "_nodelay"
external _notimeout : unit -> unit = "_notimeout" "_notimeout"
external _doupdate : unit -> unit = "_doupdate" "_doupdate"
external _printw : int -> unit = "_printw" "_printw"

type button_states = Button1Pressed | Button1Released

external has_button_state : button_states -> int -> bool
  = "has_button_state" "has_button_state"

File: lib/ncurses.c
#include <curses.h>
#include <caml/mlvalues.h>
#include <caml/memory.h>
#include <caml/alloc.h>
#include <caml/fail.h>
#include <string.h>

CAMLprim value _initscr() {
  CAMLparam0();
  CAMLlocal1(window);
  window = caml_alloc(1, Abstract_tag);
  WINDOW* window_ptr = initscr();
  if (window_ptr == NULL) caml_failwith("initscr failed");
  *((WINDOW***)Data_abstract_val(window)) = malloc(sizeof(WINDOW*));
  memcpy(*((WINDOW***)Data_abstract_val(window)), &window_ptr, sizeof(WINDOW*));
  CAMLreturn(window);
}
CAMLprim value _cbreak() {
  CAMLparam0();
  if (cbreak() == ERR) caml_failwith("cbreak failed");
  CAMLreturn(Val_unit);
}
CAMLprim value _noecho() {
  CAMLparam0();
  if (noecho() == ERR) caml_failwith("noecho failed");
  CAMLreturn(Val_unit);
}
CAMLprim value _clear() {
  CAMLparam0();
  if (clear() == ERR) caml_failwith("erase failed");
  CAMLreturn(Val_unit);
}
CAMLprim value _refresh() {
  CAMLparam0();
  if (refresh() == ERR) caml_failwith("refresh failed");
  CAMLreturn(Val_unit);
}
CAMLprim value _endwin() {
  CAMLparam0();
  if (endwin() == ERR) caml_failwith("endwin failed");
  CAMLreturn(Val_unit);
}
CAMLprim value _move(value x, value y) {
  CAMLparam2(x, y);
  int _x = Int_val(x);
  int _y = Int_val(y);
  if (move(_y, _x) == ERR) caml_failwith("move failed");
  CAMLreturn(Val_unit);
}
CAMLprim value _get_mouse() {
  CAMLparam0();
  CAMLlocal1(m_event);
  m_event = caml_alloc(3, 0);
  MEVENT m;
  mousemask(BUTTON1_PRESSED | BUTTON1_RELEASED, NULL);
  if (!has_mouse()) caml_failwith("mouse driver not initialized");
  if (getmouse(&m) == ERR) CAMLreturn(Val_none);
  Store_field(m_event, 0, Val_int(m.x));
  Store_field(m_event, 1, Val_int(m.y));
  Store_field(m_event, 2, Val_long(m.bstate));
  CAMLreturn(caml_alloc_some(m_event));
}
